Index: qemu-1.4.0/logging.h
===================================================================
--- qemu-1.4.0.orig/logging.h
+++ qemu-1.4.0/logging.h
@@ -4,6 +4,7 @@
 # include <stdarg.h>
 # include <stdio.h>
 
+/*
 #ifdef printf
 # undef printf
 #endif
@@ -16,7 +17,7 @@
 
 # define printf(...) qemu_log_printf(__VA_ARGS__)
 # define vfprintf(...) qemu_log_vfprintf(__VA_ARGS__)
-# define fprintf(...) qemu_log_fprintf(__VA_ARGS__)
+# define fprintf(...) qemu_log_fprintf(__VA_ARGS__)*/
 
 void logging_set_prefix(const char *ident);
 int qemu_log_vfprintf(FILE *stream, const char *format, va_list ap);
