diff --git a/recipes-extended/xen/xen-tools.bb b/recipes-extended/xen/xen-tools.bb
index fe0e702..7790c28 100644
--- a/recipes-extended/xen/xen-tools.bb
+++ b/recipes-extended/xen/xen-tools.bb
@@ -81,6 +81,9 @@ do_compile() {
         oe_runmake -C tools subdir-all-xenstat
         oe_runmake -C tools subdir-all-hvm-info
         oe_runmake -C tools subdir-all-xen-libhvm
+        oe_runmake -C tools subdir-all-libaio
+        oe_runmake -C tools subdir-all-blktap2
+        oe_runmake -C tools subdir-all-libxl
 }
 
 do_install() {
@@ -96,6 +99,8 @@ do_install() {
         oe_runmake DESTDIR=${D} -C tools subdir-install-xenstat
         oe_runmake DESTDIR=${D} -C tools subdir-install-hvm-info
         oe_runmake DESTDIR=${D} -C tools subdir-install-xen-libhvm
+        oe_runmake DESTDIR=${D} -C tools subdir-install-blktap2
+        oe_runmake DESTDIR=${D} -C tools subdir-install-libxl
 
 # Should not be necessary anymore
         rm -rf ${D}/etc/udev
